### DISCUSSION TITLE
How can I protect my SQL database?

### DISCUSSION THREAD

**Discussion by ghost** [community/community#123141] [Programming Help]
### Body

I got this reply from Quora:
You can’t and…

A Python programmer, C++ programmer, frontend web developer and an assembly developer for the x86 architecture. However, I am not that good at C++ , so I prefer not to get questions about C++, but you can ask me questions about HTML, Python, CSS or JavaScript.
NO YOU ARE NOT!

So do I just give up?

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by CopperEagle** [comment]
Not quite sure I understood correctly.

If you want to prevent others from using your database file, you can encrypt it. [Here](https://devconnected.com/how-to-encrypt-file-on-linux/) is a guide to do this.

Or are you looking to implement access control for a database? That is, do you want different users to see different data output? Then, you could make a "permission" field in your DB. This would be an int that encodes the level of permission a user needs to see this row. Then you can filter when querying accordingly, using e.g. `WHERE permission >= 10` in your SQL query.

If you have any fallow up questions, let me know.

**Reply by ghost** [reply]
But how can I if the first guy I asked said I cannot?

**Reply by CopperEagle** [reply]
Do you have the link for the quora discussion? There is too little context otherwhise.

**Reply by ghost** [reply]
https://www.quora.com/How-can-I-have-my-whole-entire-website-s-code-hosted-on-GitHub-but-ensure-my-database-and-server-is-secure/answer/Rob-James-179?__filter__=all&__nsrc__=notif_page&__sncid__=52086631750&__snid3__=69802058769 I have school work to do now so I won’t be able to reply.

**Reply by CopperEagle** [reply]
I see. First off, Github does not allow such (dynamic) hosting, only static html pages. 

The usual way is to host somewhere else and have all code on Github. Instead of storing the DB in the repository, you don't store any. The code generates a fresh DB during the first run. If you need access to the production DB you can log into the server and send it somewhere else.

But let's say you host on glitch.com, where repositories are typically public, including the DB. *Usually* this won't fly, indeed. The DBMS will need to write to disk without encryption. **This does not mean it is completely impossible** - at least not for glitch.com.

What you could do is having the DB purely in RAM, without a file on disk. In Python, you can do this using `sqlite3`. You could then create [regular backups](https://stackoverflow.com/questions/23395888/pure-python-backup-of-sqlite3-in-memory-database-to-disk) and encrypt these. I think you can even do it without storing to disk using `iterdump`. You can store the keys for encryption in the `.env` file, which should not be visible publicly. The drawbacks are obvious - you may loose some data upon crash and the size of your DB is limited by the available RAM.

If this answers your question, please consider marking it as the answer.

**Comment by yakir100** [comment]
There are a few ways you can do this, You can encrypt your database and it is going to protect your data from unauthorized access. Tools such as Microsoft SQL Server Transparent Data Encryption or other encryption software can help with this, You can create rules to control who can view and modify data in your database. For example, you can use SQL Server licenses to restrict what different roles can do.
